### PR TITLE
exhaustiveness: Explain why a given pattern is considered unreachable

### LINKED ISSUE #127870: unreachable pattern lint should report some (or all) of the earlier match arms that subsume the unreachable one
### Code

```Rust
enum Sym { A(i32, i32), B, C }

fn foo(s: Sym) -> &'static str {
    match s {
        Sym::B    => "B",
        Sym::A(_, 0) => "Am,0",
        Sym::A(0, _) => "A0,n",
        /* imagine a *lot* of other match arms inserted in here */
        Sym::C       => "C",
        Sym::A(0, 0) => "A0,0",
        Sym::A(_, _) => "Am,n",
    }
}

fn main() {
    foo(Sym::A(0, 0));
    foo(Sym::A(1, 1));
    foo(Sym::B);
    foo(Sym::C);
}
```

### Current output

```Shell
warning: unreachable pattern
  --> src/main.rs:10:9
   |
10 |         Sym::A(0, 0) => "A0,0",
   |         ^^^^^^^^^^^^
   |
   = note: `#[warn(unreachable_patterns)]` on by default
```

### Desired output

```Shell
warning: unreachable pattern
  --> src/main.rs:10:9
   |
10 |         Sym::A(0, 0) => "A0,0",
   |         ^^^^^^^^^^^^
   |
   = note: `#[warn(unreachable_patterns)]` on by default
   |
   = note: the first pattern that makes this pattern unreachable is:
  --> src/main.rs:6:9
   |
 6 |         Sym::A(_, 0) => "Am,0",
   |         ^^^^^^^^^^^^
```

### Rationale and extra context

I was recently adding a new enum variant that had a ton of other enum variants. I had to update an associated match expression to include it. Then, at some later point, I forgot that I had updated that match expression, redundantly added new match arms for the same variant, and got the above diangostic (which was set to be denied in my project and so it was a hard error).

I spent a little while puzzling over how the pattern could possibly be unreachable before I realized my mistake, in part because the list was so long that I had inserted the new variants at different points in the (unsorted) series of match arms. It was relatively trivial for me to diagnose in my own case, but I think my example above helps illustrate hypothetical scenarios where the compiler could do this mechanical work for us.

(Another options might be to report *all* the arms that would each independently subsume the unreachable one, as I have two such arms in my example above. But I'm not sure it would be worth the effort of computing that set ... its probably good enough to give the user the first counter-example and let them iteratively invoke the compiler if that's the usage pattern they want to follow.)

### Other cases

_No response_

### Rust Version

```Shell
playground nightly version: 1.81.0-nightly

(2024-07-12 c6727fc9b5c64cefa726)
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Nadrieril**
This PR tells the user why a given pattern is considered unreachable. I reused the intersection information we were already computing; even though it's incomplete I convinced myself that it is sufficient to always get a set of patterns that cover the unreachable one.

I'm not a fan of the diagnostic messages I came up with, I'm open to suggestions.

Fixes https://github.com/rust-lang/rust/issues/127870. This is also the other one of the two diagnostic improvements I wanted to do before https://github.com/rust-lang/rust/pull/122792.

Note: the first commit is an unrelated drive-by tweak.

r? @compiler-errors

### CODE REVIEW

File: tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr [hunk 1]
```diff
@@ -1,8 +1,10 @@
 error: unreachable pattern
   --> $DIR/exhaustiveness-unreachable-pattern.rs:8:9
    |
+LL |         (1 | 2,) => {}
+   |         -------- matches all the values already
 LL |         (1,) => {}
-   |         ^^^^
+   |         ^^^^ unreachable pattern
```
Comment (Nadrieril) [inline tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:7]: @estebank do you have a better idea for the message here? I want to express that the pattern above already matches all the values matched by the pattern below.

File: tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr [hunk 2]
```diff
@@ -13,212 +15,310 @@ LL | #![deny(unreachable_patterns)]
...
 error: unreachable pattern
   --> $DIR/exhaustiveness-unreachable-pattern.rs:19:9
    |
+LL |         (1,) => {}
+   |         ---- matches some of the same values
+LL |         (2,) => {}
+   |         ---- matches some of the same values
 LL |         (1 | 2,) => {}
-   |         ^^^^^^^^
+   |         ^^^^^^^^ unreachable pattern
```
Comment (Nadrieril) [inline tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:31]: Here I'd want to express "these patterns collectively make the last one unreachable" but I don't want to do a separate block with all the covering patterns: I like the readability of having the spans directly above the unreachable pattern span.

**Comment by Nadrieril** [comment]
Turns out we can always get a covering set of patterns!

File: compiler/rustc_pattern_analysis/tests/intersection.rs [hunk 3]
```diff
@@ -77,4 +77,14 @@ fn test_nested() {
...
+    assert_intersects(
+        pats!(ty;
+            (true, _, _),
+            (_, true, 0..10),
+            (_, true, 10..),
+            (_, true, 3),
+            _,
+        ),
+        &[&[], &[], &[], &[], &[0]],
+    );
```
Comment (Nadrieril) [inline compiler/rustc_pattern_analysis/tests/intersection.rs:89]: Something is wrong here, we should get more intersections

File: tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr (hunk 2 above)
Comment (traviscross) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:31]: Is what we do for cycle errors perhaps a useful model?  We highlight the bound that we can't check, and all of the steps of the cycle, and then say, "thus completing the cycle", etc.  This feels similar.

Comment (Nadrieril) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:31]: I do see the similarity, how do you imagine this would look like in this case?

Comment (Nadrieril) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:31]: Is this what you had in mind?
```
error: unreachable pattern
  --> $DIR/exhaustiveness-unreachable-pattern.rs:19:9
   |
LL |         (1 | 2,) => {}
   |         ^^^^^^^^ unreachable pattern
   |
note: these patterns collectively make the last one unreachable:
   |
LL |         (1,) => {}
   |         ---- matches some of the same values
LL |         (2,) => {}
   |         ---- matches some of the same values
LL |         (1 | 2,) => {}
   |         ^^^^^^^^ collectively making this unreachable
```
I like how it flows. I have a preference for not duplicating the main pattern if we can though.

Comment (traviscross) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:31]: Probably, e.g.:

```
error: unreachable pattern
  --> $DIR/exhaustiveness-unreachable-pattern.rs:19:9
   |
LL |         (1 | 2,) => {}
   |         ^^^^^^^^ unreachable pattern
   |
note: these earlier patterns collectively make the later pattern unreachable:
   |
LL |         (1,) => {}
   |         ---- matches some of the same values
LL |         (2,) => {}
   |         ---- matches some of the same values
```

Comment (Nadrieril) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:31]: I find it important for legibility that the covering patterns are shown before the unreachable one. I'd rather duplicate the unreachable one if we can't find a better phrasing

Comment (traviscross) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:31]: Seems OK to me also to duplicate it.

Comment (Nadrieril) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr]: I did the one that duplicates, ty for the idea

File: compiler/rustc_pattern_analysis/src/pat.rs [hunk 4]
```diff
@@ -147,6 +147,21 @@ impl<Cx: PatCx> fmt::Debug for DeconstructedPat<Cx> {
...
+    }
+}
+/// Delegate to `uid`.
+impl<Cx: PatCx> Eq for DeconstructedPat<Cx> {}
+/// Delegate to `uid`.
+impl<Cx: PatCx> std::hash::Hash for DeconstructedPat<Cx> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.uid.hash(state);
+    }
+}
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/pat.rs:163]: Because of `covered_by: FxHashSet<&'p DeconstructedPat<Cx>>`. There's no map from uids to patterns so I can't store just the uids. I could have used the derived `Hash`, it would just have been much worse perf.

[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_pattern_analysis/src/pat.rs:163]: Yeah ok, just an optimization is what I was curious about.

File: tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr (hunk 1 above)
Comment (estebank) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:7]: What about "unreachable pattern because it is already fully covered by a(n) earlier one(s)"?

Comment (compiler-errors) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:7]: or slightly shorter: "unreachable pattern is already fully covered by an earlier one"

File: tests/ui/pattern/usefulness/empty-types.min_exh_pats.stderr [hunk 5]
```diff
@@ -403,18 +456,24 @@ error: unreachable pattern
    |
 LL |         _ => {}
    |         ^
+   |
+   = note: this pattern matches no values because `[!; 3]` is uninhabited
```
Comment (estebank) [inline tests/ui/pattern/usefulness/empty-types.min_exh_pats.stderr:460]: We should look into making this the span label and maybe have the note be a link to an explanation of uninhabited types (!).

File: tests/ui/pattern/usefulness/explain-unreachable-pats.stderr [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+LL |         | 0, _) = (0, 0) {}
+   |           ^ unreachable pattern
+
+error: unreachable pattern
+  --> $DIR/explain-unreachable-pats.rs:71:9
+   |
+LL |         (_, true) => {}
+   |         ^^^^^^^^^ unreachable pattern
+   |
+note: these patterns collectively make the last one unreachable
```
Comment (estebank) [inline tests/ui/pattern/usefulness/explain-unreachable-pats.stderr:73]: We're not assured that the unreachable pattern was the last one, so we should tweak the wording of this note message to be something along the lines "these patterns collectively make the other one unreachable" or something.

**Comment by estebank** [comment]
@Nadrieril thank you for doing this! It's gonna be super helpful.

File: tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr (hunk 1 above)
Comment (Nadrieril) [reply tests/ui/or-patterns/exhaustiveness-unreachable-pattern.stderr:7]: I like this. What about the "matches all the values already"?

File: tests/ui/error-codes/E0001.stderr [hunk 7]
```diff
@@ -2,8 +2,17 @@ error: unreachable pattern
...
    |
+note: these patterns collectively make the last one unreachable
+  --> $DIR/E0001.rs:8:9
+   |
+LL |         Some(_) => {/* ... */}
+   |         ------- matches some of the same values
+LL |         None => {/* ... */}
+   |         ---- matches some of the same values
+LL |         _ => {/* ... */}
+   |         ^ collectively making this unreachable
```
Comment (tgross35) [inline tests/ui/error-codes/E0001.stderr:15]: I find the flat line a bit hard to parse, looks more like a separator or a leader to another arrow than an indicator itself. Any specific reason this uses the spatula rather than the forks?

(regardless, this change is wonderful)

**Comment by Nadrieril** [comment]
@estebank let's chat about this, I can make another PR to tweak the messages

**Comment by Nadrieril** [comment]
Is this worthy of relnotes? I think it's cute

File: tests/ui/error-codes/E0001.stderr (hunk 7 above)
Comment (Nadrieril) [reply tests/ui/error-codes/E0001.stderr:15]: I don't know what "spatula" and "forks" mean, I think just used the default principal/non-principal span mechanism in diagnostics

Comment (tgross35) [reply tests/ui/error-codes/E0001.stderr:15]: I was being too creative, but

`______--` spatula

`^^---` fork

I figured it wasn't anything done specifically here. This diagnostic just winds up with more alternating lines of code and `----` than many others, and I think it is somewhat ambiguous whether `----` refers to the line up or down.

Obviously not a big deal since it is just a hint anyway.

Comment (Nadrieril) [reply tests/ui/error-codes/E0001.stderr:15]: I see. It looks a lot clearer with color at least

File: tests/ui/pattern/usefulness/explain-unreachable-pats.stderr (hunk 6 above)
Comment (Nadrieril) [reply tests/ui/pattern/usefulness/explain-unreachable-pats.stderr:73]: Hm, it may not be the last one in temrs of spans, but it has to be the last one in terms of match order
